python-lib/dku_deeplearning_image/utils.py

```diff
@@ -164,33 +171,75 @@ def wrapper(*args, **kwargs):
+                    [(id_pred(i), float(pred[i])) for i in pred.argsort()[-limit:] if float(pred[i]) >= min_threshold])
+                formatted_predictions.append(json.dumps(formatted_pred))
+            else:
+                formatted_predictions.append(pred.tolist())
+        else:
+            logger.warning(f"There has been an error with prediction: {pred_i}. (It is probably not an image)")
+            formatted_predictions.append(None)
+    return {"prediction": formatted_predictions, "error": errors}
+
 
+def apply_preprocess_image(tfds, input_shape, preprocessing, is_b64=False):
+    def _apply_preprocess_image(image_path):
+        return tf.numpy_function(
+            func=lambda x: preprocess_img(x, input_shape, preprocessing, is_b64),
+            inp=[image_path],
+            Tout=tf.float32)
 
-def get_predictions(model, batch, classify=False, limit=constants.DEFAULT_PRED_LIMIT, min_threshold=0, labels_df=None):
-    predictions = model.predict(batch)
-    return format_predictions_output(predictions, classify, labels_df, limit, min_threshold)
+    def _convert_errors(images):
+        return tf.numpy_function(
+            func=lambda x: tf.cast(x.size == 0, tf.int8),
+            inp=[images],
+            Tout=tf.int8)
 
+    def _filter_errors(images):
+        return tf.numpy_function(
+            func=lambda x: x.size != 0,
+            inp=[images],
+            Tout=tf.bool)
 
-def get_ordered_dict(predictions):
-    return json.dumps(OrderedDict(sorted(predictions.items(), key=(lambda x: -x[1]))))
+    preprocessed_images = tfds.map(map_func=_apply_preprocess_image, num_parallel_calls=constants.AUTOTUNE)
+    error_array = preprocessed_images.map(map_func=_convert_errors, num_parallel_calls=constants.AUTOTUNE)
+    preprocessed_images_filtered = preprocessed_images.filter(predicate=_filter_errors)
 
+    return preprocessed_images_filtered, error_array
 
-def preprocess_img(img_path, img_shape, preprocessing):
+
+def read_images_to_tfds(images_folder, np_images):
+    def retrieve_image_from_folder(image_fn):
+        return tf.numpy_function(
+            func=lambda x: get_cached_file_from_folder(images_folder, x),
+            inp=[image_fn],
+            Tout=tf.string)
+
+    X_tfds = tf.data.Dataset.from_tensor_slices(np_images.reshape(-1, 1))
+    return X_tfds.interleave(
+        map_func=lambda x: tf.data.Dataset.from_tensor_slices(x).map(
+            retrieve_image_from_folder, num_parallel_calls=constants.AUTOTUNE),
+        num_parallel_calls=constants.AUTOTUNE)
+
+
+def preprocess_img(img_path, img_shape, preprocessing, is_b64=False):
     try:
+        if is_b64:
+            img_path = BytesIO(img_path)
         img = Image.open(img_path).resize(img_shape[:2])
     except UnidentifiedImageError as err:
         logger.warning(f'The file {img_path} is not a valid image. skipping it. Error: {err}')
-        return
+        return tf.cast(np.array([]), tf.float32)
```

Comment:
You are returning a tf tensor here, while if no error, you are returning a numpy array. I don't know if this works, but definitely does not look clean

Reply:
1. Yop, fixed in 0f2e0cd
2. True, i was a real pain to build this function, and the return value if there is an error but you are right, things should be consistant. Fixed in b6310a2

Reply:
I think that `preprocess_img` should return only `np.ndarray` and not `tf.tensor`. The goal of [`numpy_function`](https://www.tensorflow.org/api_docs/python/tf/numpy_function) is to abstract away the complexity of tensorflow from the used function. Tensorflow is in charge of converting the output to tensors

`preprocess_img` could be used outside of tensorflow, but currently it's not possible

Reply:
I did the casting outside the function in 287bfe7, thanks!